wagtail/wagtailcore/models.py

```diff
@@ -1277,17 +1283,184 @@ def serve_password_required_response(self, request, form, action_url):
+    administer all children of those paths, too).
+    The second list is the paths which must be shown to the user in order for them to navigate to their permitted Pages
+    in the Explorer. Unlike the paths in the first list, the children of these paths are not to be automatically made
+    visible.
+
+    NOTE: The first element of the second list is always the Closest Common Ancestor of the User's permitted Pages.
+    """
+    # We cache the output of this function because it's very expensive to generate it fresh every time the user loads
+    # an Explorer page. Unfortunately, this does lead to a lot of duplciated data in the cache, as every single user
+    # that has the same set of groups will have the same permitted paths. I'd like to cache by groupset, which would
+    # remove all that duplication, but that would require an extra DB call every time, to get the user's group list.
+    cache_key = EXPLORABLE_PATHS_PREFIX.format(user.pk)
+    permitted_paths, required_ancestors = cache.get(cache_key, ([], []))
+    if not permitted_paths:
+        ###################
+        # Get paths by User
+        ###################
+        permissions = (
+            GroupPagePermission.objects
+            .filter(group__user=user)
+            # Choose permission is excluded because it's only associated with the Choosers, not the Explorer.
+            .exclude(permission_type='choose')
+            .select_related('page')
+        )
+        # We temporarily use a set here because "permissions" has duplicates of every path that has multiple perms.
+        permitted_paths = list(set(perm.page.path for perm in permissions))
+
+        # Calculate the "closest common ancestor" of all the permitted Pages by finding their common prefix, then
+        # chopping off the end to make the length a multiple of Page.steplen.
+        if permitted_paths:
+            common_prefix = os.path.commonprefix(permitted_paths)
+            cca_path = common_prefix[:-(len(common_prefix) % Page.steplen) or None]
+            required_ancestors.append(cca_path)
+            # Now include all the ancestors of each permitted Page, up to their closest common ancestor. This lets
+            # users see parent pages that they can't administer, which allows them to traverse the tree down to pages
+            # that they can.
+            temp = set()
+            for path in permitted_paths:
+                while len(path) > len(cca_path):
+                    path = path[:-Page.steplen]
+                    if path not in required_ancestors:
+                        temp.add(path)
+            # We need to make sure that the CCA is first in required_ancestors (so the /admin/pages/ page can be set to
+            # display the page tree starting at the CCA), but the order of the other paths doesn't matter.
+            required_ancestors.extend(temp)
```

Comment:
Do we really need to find all `required_ancestors`? I'm not sure, but closest common ancestor  probably will be enough. See:
* [Here](https://github.com/torchbox/wagtail/pull/2463/files#diff-d099bfc6462f8efe8cc5177b2790e900R49), [here](https://github.com/torchbox/wagtail/pull/2463/files#diff-9ec0dccfdaa33a0fd310c76b5db2cb37R1328) and [here](https://github.com/torchbox/wagtail/pull/2463/files#diff-9ec0dccfdaa33a0fd310c76b5db2cb37R1448) we can to use only closest common ancestor, if present
* [Here](https://github.com/torchbox/wagtail/pull/2463/files#diff-9ec0dccfdaa33a0fd310c76b5db2cb37R1297) we can use only `permitted_paths` (`path__startswith` because it's not possible to have a subpage with less permissions than the parent)
* [Here](https://github.com/torchbox/wagtail/pull/2463/files#diff-9ec0dccfdaa33a0fd310c76b5db2cb37R1318) we need to find all paths in range from "closest common ancestor" to `self.path`

Please, correct me if I'm wrong.